include/albatross/src/tune/greedy_tuner.hpp

```diff
@@ -0,0 +1,208 @@
+ *   - Using the range defined by the prior to produce new guesses.
+ *   - Evaluating the objective at all the guesses (one guess per thread).
+ *   - Picking the best parameter value.
+ *   - Repeating for the next parameter.
+ *
+ * There are almost certainly no global convergence guarantees with
+ * this algorithm, but in practice it seems to do a good job of coarsely
+ * tuning the objective.  One strategy would be to start with an iteration
+ * of this greedy approach and follow up with one of the methods in tune.hpp
+ */
+
+namespace details {
+
+inline std::vector<double> query_ratios(std::size_t n) {
+  const double log_min = -5.;
+  const double log_max = -0.1;
+  if (n == 1) {
+    return {log_min};
+  }
+
+  const double step = (log_max - log_min) / (n - 1);
+  std::vector<double> output;
+  for (std::size_t i = 0; i < n; ++i) {
+    output.push_back(std::pow(10, log_min + i * step));
+  }
+
+  return output;
+}
+
+inline std::vector<double> get_queries(double value, double low, double high,
+                                       std::size_t n) {
+
+  if (std::isinf(high)) {
+    high = 1e8;
+  }
+  if (std::isinf(low)) {
+    low = -1e8;
+  }
+
+  const double low_range = value - low;
+  const double high_range = high - value;
+
+  std::vector<double> queries;
+
+  if (value == low) {
+    // only search higher since we're at the lower bound
+    const auto ratios = query_ratios(2 * n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value + high_range * ratios[i]);
+    }
+  } else if (value == high) {
+    // only search lower since we're at the upper bound
+    const auto ratios = query_ratios(2 * n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value - low_range * ratios[ratios.size() - 1 - i]);
+    }
+  } else {
+    // search in both directions
+    const auto ratios = query_ratios(n);
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value - low_range * ratios[ratios.size() - 1 - i]);
+    }
+    for (std::size_t i = 0; i < ratios.size(); ++i) {
+      queries.push_back(value + high_range * ratios[i]);
+    }
+  }
+
+  return queries;
+}
+
+inline ParameterStore set_tunable_param(const ParameterStore &params,
+                                        std::size_t i, double val) {
+  auto perturbed = get_tunable_parameters(params);
+  perturbed.values[i] = val;
+  return set_tunable_params_values(params, perturbed.values, true);
+};
+} // namespace details
+
+template <typename Function>
+inline ParameterStore
+greedy_tune(Function evaluate_function, const ParameterStore &params,
+            std::size_t n_threads = 8, std::size_t n_iterations = 10,
+            bool use_async = true, std::ostream *os = &std::cout) {
+
+  static_assert(
+      has_call_operator<Function, ParameterStore>::value,
+      "evaluate_function must have a single ParameterStore argument.");
+
+  assert(n_threads % 2 == 0); // n_threads must be even;
```

Comment:
why does n_threads have to be even?

Reply:
The algorithm splits the domain to be searched into "below" and "above" the current value and I allocate an equal number of threads to each.